frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py

```diff
@@ -0,0 +1,851 @@
+    # We know from the definition of qalloc_p
+    # that there is only one operand
+    # and the operand is the size of the register.
+    size = invals[0]
+    register = kernel_qalloc(kernel, size)
+    outvals = [register]
+
+    # We are creating a new variable that will replace the old
+    # variable.
+    outvariables = [ctx.new_variable(AbsCudaQReg())]
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, outvals)
+    return register
+
+
+def change_register_getitem(ctx, eqn):
+    """Change catalyst's qextract_p primitive to a CUDA-quantum primitive."""
+
+    assert eqn.primitive == qextract_p
+    invals = safe_map(ctx.read, eqn.invars)
+    # We know from the definition of qextract_p
+    # that it takes two operands.
+    # The first one is the qreg and the second one is the
+    # qubit index.
+    # Because we have correctly mapped the replacement,
+    # invals[0] should point to a correct cuda register.
+    register = invals[0]
+    idx = invals[1]
+    cuda_qubit = qreg_getitem(register, idx)
+    ctx.set_qubit_to_wire(idx, cuda_qubit)
+
+    outvariables = [ctx.new_variable(AbsCudaQbit())]
+    outvals = [cuda_qubit]
+
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, outvals)
+
+
+def change_register_setitem(ctx, eqn):
+    """Set the correct post-conditions for CUDA-quantum when interpreting qinsert_p primitive
+
+    This method is interesting because CUDA-quantum does not use value semantics for their qubits.
+    This means that each qubit is transformed as a side effect of the operations. Since
+    CUDA-quantum does not use value semantics for their qubits nor their quantum registers,
+    it means that we must map from a value semantics program to a memory semantics program.
+
+    How do we do that?
+    We remove the extra SSA variables, but keep track of which variables refer to which qubits.
+    """
+
+    # There is no __setitem__ for a quake value.
+    assert eqn.primitive == qinsert_p
+    # We know from the definition of qinsert_p
+    # that it takes three operands
+    # * qreg_old
+    # * idx
+    # * qubit
+    #
+    # qinsert_p also returns a new qreg
+    # which represents the new state of the previous
+    # qreg after the insertion.
+    # * qreg_new
+    #
+    # We are just going to map the new register to
+    # the old register, and I think that will keep
+    # the same semantics in place.
+    invals = safe_map(ctx.read, eqn.invars)
+    # Because invals has been replaced with the correct
+    # variables, invals[0] now holds a reference to a cuda register
+    old_register = invals[0]
+    # This is old we need to do.
+    safe_map(ctx.write, eqn.outvars, [old_register])
+
+
+def change_instruction(ctx, eqn):
+    """Change the instruction to one supported in CUDA-quantum."""
+
+    assert eqn.primitive == qinst_p
+
+    # This is the map of instruction names.
+    from_catalyst_to_cuda = {
```

Comment:
could we add this as a static class variable?

Reply:
Can we keep it? It is used immediately there. I like to keep variables close to their uses.